numpy/lib/function_base.py

```diff
@@ -3929,21 +3928,33 @@ def _quantile_ureduce_func(a, q, axis=None, out=None, overwrite_input=False,
             indices_above = indices_above[:-1]
             n = np.isnan(ap[-1:, ...])
 
-        x1 = take(ap, indices_below, axis=axis) * weights_below
-        x2 = take(ap, indices_above, axis=axis) * weights_above
+        x1 = take(ap, indices_below, axis=axis)
+        x2 = take(ap, indices_above, axis=axis)
+
+        # the linear interpolation below ensure monotonicity with floating
+        # point operations and was proposed in:
+        # https://math.stackexchange.com/a/1798323 by Pedro Gimeno
+        diff_x2_x1 = x2 - x1
 
-        # ensure axis with q-th is first
         x1 = np.moveaxis(x1, axis, 0)
         x2 = np.moveaxis(x2, axis, 0)
```

Comment:
Move this before the `diff` calcluation I think.

Reply:
You'll need to call moveaxis on the weight too, I think

Reply:
Wait, do we even need this? We already moved the axis above...

I think perhaps the take should be `axis=0`

Reply:
Ah, axis is already 0 here, so this is a no-op